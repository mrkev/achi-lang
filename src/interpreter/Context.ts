import { Value, NamedRecordConstructor } from "./interpreter";
import { nullthrows } from "./nullthrows";

/**
 * Context handles variable/ scoping, stack frames (TODO), etc
 */
export class Context {
  static create() {
    return new Context();
  }

  private valueScopes: Scope<Value>[] = [new Scope(null)];

  // TODO: standard types, like records?
  types: Map<string, NamedRecordConstructor> = new Map();

  pushScope() {
    const currentScope = this.valueScopes.at(-1);
    if (!currentScope) {
      throw new Error("No current scope");
    }
    this.valueScopes.push(currentScope);
  }

  popScope() {
    if (this.valueScopes.length === 1) {
      throw new Error("Attempting to pop global (frist) scope!");
    }
    this.valueScopes.pop();
  }

  values(): Scope<Value> {
    const currentScope = nullthrows(this.valueScopes.at(-1));
    return currentScope;
  }
}

class Scope<T> {
  parent: Scope<T> | null;
  map: Map<string, T> = new Map();

  constructor(parent: Scope<T> | null) {
    this.parent = parent;
  }

  has(identifer: string): boolean {
    if (this.map.has(identifer)) {
      return true;
    } else if (this.parent != null) {
      return this.parent.has(identifer);
    } else {
      return false;
    }
  }

  define(identifer: string, value: T) {
    // only disallow overriding own scope
    if (this.map.has(identifer)) {
      throw new Error(`${identifer} is already defined`);
    }
    this.map.set(identifer, value);
  }

  // TODO: override in its right scope
  set(identifer: string, value: T) {
    if (!this.has(identifer)) {
      throw new Error(`no definition for ${identifer} found`);
    }
    this.map.set(identifer, value);
  }

  get(identifer: string): T | null {
    if (this.has(identifer)) {
      return this.get(identifer);
    }
    if (this.parent != null) {
      return this.parent.get(identifer);
    }
    return null;
  }
}
