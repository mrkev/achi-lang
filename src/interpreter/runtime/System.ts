import { ScriptError } from "../interpreterErrors";

/**
 * System handles I/O, other APIs
 */
export class System {
  readonly console: Console = new Console();
}

class Console {
  _log: (string | Error)[] = [];
  _fatalError: Error | ScriptError | null = null;

  clear() {
    this._log = [];
  }

  log(...msgs: string[]) {
    this._log.push(...msgs);
  }

  fail(err: Error) {
    this._fatalError = err;
    this._log.push(err);
  }

  output(): string {
    return this._log.join("\n");
  }
}
