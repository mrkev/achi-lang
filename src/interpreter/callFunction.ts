import { Context } from "./Context";
import {
  ReturnInterrupt,
  evaluateStatements,
  evaluateWithScope,
} from "./evaluateStatements";
import { exhaustive, nullthrows } from "../nullthrows";
import { System } from "./runtime/System";
import { ValueType, nil } from "./runtime/value";
import { destructureWithRecordDefintion } from "./evaluateExpression";
import { dynamicTypecheck } from "./dynamicTypechecker";

// // The type of destructuring for assignments, without types
// // const (a: foo, b) = thing;
// function destructureWithRecordLiteral() {
//   // empty
// }
export function callFunction(
  func:
    | ValueType["MatchFunctionInstance"]
    | ValueType["AnonymousFunctionInstance"],
  argument: ValueType["Value"],
  context: Context,
  system: System
): ValueType["Value"] {
  switch (func.kind) {
    case "AnonymousFunctionInstance": {
      context.stack.push(func);
      // console.log("called");

      // TODO: check if incoming record conforms to argument.
      dynamicTypecheck(argument, func.ast.argument);

      // If not, error will show in `destructureWithRecordDefintion` as the
      // destructuring failing, which is confusing
      const returned = evaluateWithScope(context, () => {
        destructureWithRecordDefintion(func.ast.argument, argument, context);
        evaluateStatements(func.ast.block.statements, context, system);
      });

      context.stack.pop();

      if (returned == null) {
        // TODO: do I really want functions to return an implicit null?
        return nil(returned);
      } else {
        return returned;
      }
    }

    case "MatchFunctionInstance": {
      // TODO: test types
      // TODO: ensure pattern is exhaustive, so we don't have to worry about that here
      // TODO: just executing first case for now
      for (const caseEntry of func.ast.block.caseEntries) {
        // caseEntry.guard
        context.valueScope.push();
        context.stack.push(func);

        let result = null;
        try {
          evaluateStatements(caseEntry.block.statements, context, system);
        } catch (interrupt) {
          if (interrupt instanceof ReturnInterrupt) {
            result = interrupt.value;
          } else {
            throw interrupt;
          }
        }

        context.stack.pop();
        context.valueScope.pop();
        return nullthrows(result, "fixme, case block returns nothing");
      }

      throw new Error("fixme, no case blocks");
    }

    default: {
      throw exhaustive(func);
    }
  }
}
