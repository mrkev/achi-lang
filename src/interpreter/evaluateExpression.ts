import { LangType } from "../parser/parser";
import { Value, NamedRecordInstance } from "./interpreter";
import { exhaustive, nullthrows } from "./nullthrows";
import { Context } from "./Context";
import { evaluateStatements } from "./evaluateStatements";
import { System } from "./System";
import { RecordLiteralInstance } from "./runtime/NamedRecordConstructor";

export function evaluateExpression(
  expression: LangType["Expression"],
  context: Context,
  system: System
): Value {
  const { kind } = expression;
  switch (kind) {
    case "ValueIdentifier": {
      const foundValue = context.values().get(expression.value);
      if (!foundValue) {
        throw new Error(`Name ${expression.value} not found`);
      }
      return foundValue;
    }

    case "NamedRecordLiteral": {
      system.console.log("->NamedRecordLiteral");
      const instance = NamedRecordInstance.fromNamedRecordLiteral(
        expression,
        context,
        system
      );
      return { kind: "NamedRecord", value: instance };
    }

    case "NumberLiteral": {
      return { kind: "number", value: expression.value };
    }

    case "RecordLiteral": {
      system.console.log("->RecordLiteral");
      const props = new Map<string, Value>
      for (const def of expression.definitions) {
        const value = evaluateExpression(def.expression, context, system)
        // TODO: ensure no duplicate identifiers
        props.set(def.identifier.value, value)
      }

      const instance = new RecordLiteralInstance(expression, props)
      return {kind: "RecordLiteralInstance", value: instance}
    }

    case "StringLiteral": {
      return { kind: "string", value: expression.value };
      // throw new Error("Not implemented; StringLiteral evaluation");
    }

    case "FunctionDefinition": {
      throw new Error("Not implemented; FunctionDefinition evaluation");
    }

    case "FunctionCall": {
      const { identifier, argument } = expression;

      const funcInstance = context
        .values()
        .getOrThrow(
          identifier.value,
          `No definition for function ${identifier.value}`
        );

      // TODO: other function types
      const instKind = funcInstance.kind;
      if (instKind !== "MatchFunctionInstance") {
        throw new Error(
          `${identifier.value} is a ${funcInstance.kind}, not a function`
        );
      }
      const argumentValue = evaluateExpression(argument, context, system);
      funcInstance.value.ast.block;

      return callFunction(
        funcInstance.value.ast,
        argumentValue,
        context,
        system
      );
    }

    default:
      throw exhaustive(kind);
  }
}

function callFunction(
  func: LangType["MatchFunction"],
  argument: Value,
  context: Context,
  system: System
): Value {
  // TODO: handle different types of function
  // TODO: test types
  // TODO: just executing first case for now

  for (const caseEntry of func.block.caseEntries) {
    // caseEntry.guard
    const result = evaluateStatements(
      caseEntry.block.statements,
      context,
      system
    );
    return nullthrows(result, "fixme, case block returns nothing");
  }

  throw new Error("fixme, no case blocks");
}
