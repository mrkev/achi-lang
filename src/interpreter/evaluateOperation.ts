import { boolean, expectBoolean, Value } from "./value";
import { Context } from "./Context";
import { System } from "./runtime/System";
import {
  BinaryOperation,
  PrefixUnaryOperation,
  SuffixUnaryOperation,
} from "../parser/parser.binex";
import { factorial } from "./runtime/utils";
import { expectNumber, expectString, number, string } from "./value";
import { evaluateExpression } from "./evaluateExpression";
import { exhaustive } from "./nullthrows";

export function evaluatePrefixUnaryExpression(
  unex: PrefixUnaryOperation,
  context: Context,
  system: System
): Value {
  const value = evaluateExpression(unex.value, context, system) as Value;

  switch (unex.operator) {
    case "-":
      return number(-expectNumber(value).value);
    case "!":
      return boolean(!expectBoolean(value).value);
    default:
      exhaustive(unex.operator);
      throw new Error("unknown operator " + unex.operator);
  }
}

export function evaluateSuffixUnaryExpression(
  unex: SuffixUnaryOperation,
  context: Context,
  system: System
): Value {
  const value = evaluateExpression(unex.value, context, system) as Value;
  switch (unex.operator) {
    case "!":
      return number(factorial(expectNumber(value).value));
    default:
      exhaustive(unex.operator);
      throw new Error("unknown operator " + unex.operator);
  }
}

export function evaluateBinaryExpression(
  binex: BinaryOperation,
  context: Context,
  system: System
): Value {
  const left = evaluateExpression(binex.left, context, system) as Value;
  const right = evaluateExpression(binex.right, context, system) as Value;

  switch (binex.operator) {
    case "*": {
      return number(expectNumber(left).value * expectNumber(right).value);
    }
    case "+": {
      switch (left.kind) {
        case "number": {
          const result = expectNumber(left).value + expectNumber(right).value;
          return number(result);
        }
        case "string": {
          const result = expectString(left).value + expectString(right).value;
          return string(result);
        }
        default:
          throw new Error("UNEXPECTED VALUE FOR BINOP " + left.kind);
      }
    }
    case "-":
      return number(expectNumber(left).value - expectNumber(right).value);
    case "^": {
      return number(
        Math.pow(expectNumber(left).value, expectNumber(right).value)
      );
    }
    case "/":
      return number(expectNumber(left).value / expectNumber(right).value);
    default:
      exhaustive(binex.operator);
      throw new Error("unknown operator " + binex.operator);
  }
}
