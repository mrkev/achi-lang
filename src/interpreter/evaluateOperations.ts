import { Value } from "./value";
import { Context } from "./Context";
import { System } from "./runtime/System";
import { BinaryOperation, UnaryOperation } from "../parser/parser.binex";
import { factorial } from "./runtime/utils";
import { expectNumber, expectString, number, string } from "./value";
import { evaluateExpression } from "./evaluateExpression";

export function evaluateUnaryExpression(
  unex: UnaryOperation,
  context: Context,
  system: System
): number {
  const value = expectNumber(
    evaluateExpression(unex.value, context, system) as Value
  ).value;

  switch (unex.operator) {
    case "-":
      return -value;
    case "!":
      return factorial(value);
    case "*":
    case "+":
    case "^":
      throw new Error("invalid unary operator " + unex.operator);
    default:
      throw new Error("unknown operator " + unex.operator);
  }
}

export function evaluateBinaryExpression(
  binex: BinaryOperation,
  context: Context,
  system: System
): Value {
  const left = evaluateExpression(binex.left, context, system) as Value;
  const right = evaluateExpression(binex.right, context, system) as Value;

  switch (binex.operator) {
    case "*": {
      return number(expectNumber(left).value * expectNumber(right).value);
    }
    case "+": {
      switch (left.kind) {
        case "number": {
          const result = expectNumber(left).value + expectNumber(right).value;
          return number(result);
        }
        case "string": {
          const result = expectString(left).value + expectString(right).value;
          return string(result);
        }
        default:
          throw new Error("UNEXPECTED VALUE FOR BINOP " + left.kind);
      }
    }
    case "-":
      return number(expectNumber(left).value - expectNumber(right).value);
    case "^": {
      return number(
        Math.pow(expectNumber(left).value, expectNumber(right).value)
      );
    }
    case "!":
      throw new Error("invalid binary operator " + binex.operator);
    default:
      throw new Error("unknown operator " + binex.operator);
  }
}
