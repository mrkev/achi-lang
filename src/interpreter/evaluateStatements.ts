import { LangType } from "../parser/parser";
import { evaluateExpression } from "./evaluateExpression";
import { Context } from "./Context";
import { System } from "./System";
import { exhaustive } from "./nullthrows";
import { stringOfValue, Value } from "./interpreter";
import { NamedRecordKlass } from "./runtime/NamedRecordConstructor";
import { MatchFunctionInstance } from "./runtime/MatchFunctionInstance";

/** Evaluates all statements sequentially, returning the value of the first "ReturnStatement" found or void */
export function evaluateStatements(
  statements: LangType["Statement"][],
  context: Context,
  system: System
): Value | null {
  for (const statement of statements) {
    const { kind } = statement;
    switch (kind) {
      case "NamedRecordDefinitionStatement": {
        const identifer = statement.namedRecordDefinition.identifier.value;
        if (
          context.types.has(statement.namedRecordDefinition.identifier.value)
        ) {
          console.warn(
            "Overriding definition for",
            statement.namedRecordDefinition.identifier.value
          );
        }
        const klass = NamedRecordKlass.fromNamedRecordDefinition(
          statement.namedRecordDefinition
        );
        // todo "define" not set (ie, error if already exists)
        context.types.set(
          statement.namedRecordDefinition.identifier.value,
          klass
        );
        context
          .values()
          .define(identifer, { kind: "NamedRecordKlass", value: klass });
        break;
      }

      case "NamedRecordDefinitionGroup": {
        console.log("TODO NamedRecordDefinitionGroup");
        break;
      }

      case "IfStatement": {
        const guardValue = evaluateExpression(statement.guard, context, system);
        // TODO: determine truthiness
        if (guardValue) {
          // TODO: scoping!
          evaluateStatements(statement.block.statements, context, system);
        }
        break;
      }

      case "ConstantAssignment": {
        if (context.values().has(statement.identifier.value)) {
          console.warn("Overriding definition for", statement.identifier.value);
        }
        const result = evaluateExpression(
          statement.expression,
          context,
          system
        );
        context.values().define(statement.identifier.value, result);
        break;
      }

      case "DEBUG_Log": {
        const value = evaluateExpression(statement.expression, context, system);
        system.console.log(stringOfValue(value));
        break;
      }

      case "ReturnStatement": {
        const value = evaluateExpression(statement.expression, context, system);
        // We don't evaluate further, and instead return immediately //
        return value;
      }

      // evaluates
      case "FunctionCall": {
        const _result = evaluateExpression(statement, context, system);
        // when functions are called as statements, we ignore the result
        break;
      }

      // defines
      case "MatchFunction": {
        if (context.values().has(statement.identifier.value)) {
          console.warn(
            "MatchFunction: Overriding definition for",
            statement.identifier.value
          );
        }

        const matchFuncInstance = new MatchFunctionInstance(statement);
        context.values().define(statement.identifier.value, {
          kind: "MatchFunctionInstance",
          value: matchFuncInstance,
        });
        break;
      }

      default:
        exhaustive(kind);
    }
  }
  return null;
}
