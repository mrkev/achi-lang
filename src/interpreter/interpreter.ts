import { LangType, tryParse } from "../parser/parser";
import { Context, ScopeError } from "./Context";
import { System } from "./runtime/System";
import { exhaustive, nullthrows } from "./nullthrows";
import { evaluateStatements } from "./evaluateStatements";
import {
  NamedRecordKlass,
  RecordLiteralInstance,
  NamedRecordDefinitionGroupInstance,
  NamedRecordInstance,
} from "./runtime/runtime.records";
import { MatchFunctionInstance } from "./runtime/runtime.match";
import { AnonymousFunctionInstance } from "./runtime/runtime.functions";
import { TypeMismatchError } from "../checker/checker";

export function interpret(
  script: string | LangType["Program"],
  system: System = new System(),
  context: Context = Context.create()
): void {
  try {
    const ast = typeof script === "string" ? tryParse(script) : script;
    context.valueScope.push();
    evaluateStatements(ast.statements, context, system);
    context.valueScope.pop();
  } catch (e) {
    if (e instanceof ScopeError) {
      if (typeof script === "string") {
        const nice = niceError(script, e);
        console.log("nice");
        system.console.log(nice);
      }
    } else if (e instanceof Error || typeof e === "string") {
      system.console.log(e);
    } else {
      console.log("Unknown error type during interpretation");
    }
    console.error(e);
  }
}

export function niceError(
  script: string,
  error: ScopeError | TypeMismatchError
) {
  const lines = script.split("\n");
  const numLineDigits = lines.length.toString().length;
  const startLine = error.location().start.line - 1; // make it 0 indexed
  const lineNum = (num: number) =>
    (num - 1).toString().padStart(numLineDigits + 1);

  const msg = [];
  if (startLine > 0) {
    msg.push(`${lineNum(startLine + 1)}| ${lines[startLine - 1]}`);
  }
  msg.push(`${lineNum(startLine + 2)}| ${lines[startLine]}`);
  msg.push(
    "".padStart(numLineDigits + 1) +
      "  " +
      "^".padStart(error.location().end.column - 1)
  );
  if (lines.length > startLine) {
    msg.push(`${lineNum(startLine + 3)}| ${lines[startLine + 1]}`);
  }
  msg.push("");
  msg.push(error.print());
  return msg.join("\n");
}

export type Value =
  // 3
  | { kind: "number"; value: number }
  // "hello"
  | { kind: "string"; value: string }
  // false
  | { kind: "boolean"; value: boolean }
  // null
  | { kind: "empty"; value: null }
  // Point(x: 3, y: 2)
  | { kind: "NamedRecordInstance"; value: NamedRecordInstance }
  // function printPoint matches (point: Point) { ... }
  | { kind: "MatchFunctionInstance"; value: MatchFunctionInstance }
  // class Point(x: number, y: number)
  | { kind: "NamedRecordKlass"; value: NamedRecordKlass }
  // (x: 3, y: 2)
  | { kind: "RecordLiteralInstance"; value: RecordLiteralInstance }
  // classes Cards { ... }
  | {
      kind: "NamedRecordDefinitionGroupInstance";
      value: NamedRecordDefinitionGroupInstance;
    }
  | {
      kind: "AnonymousFunctionInstance";
      value: AnonymousFunctionInstance;
    };

export function stringOfValue(value: Value): string {
  const { kind } = value;
  switch (kind) {
    case "number":
    case "string":
    case "boolean": {
      return String(value.value);
    }
    case "NamedRecordInstance": {
      let str = value.value.konstructor.classname + " {";
      const entries = [...value.value.recordLiteral.props.entries()];
      if (entries.length > 0) {
        str += "\n";
      }
      for (const [key, val] of entries) {
        str += `  ${key}: ${stringOfValue(
          val
        )} (${value.value.konstructor.valueSpec.get(key)})\n`;
      }
      str += "}";
      return str;
    }

    case "MatchFunctionInstance": {
      return `[Function: ${value.value.identifier}]`;
    }

    case "NamedRecordKlass": {
      return `[Class: ${value.value.classname}]`;
    }

    case "RecordLiteralInstance": {
      // return "todo RecordLiteral";
      let str = "(\n";
      for (const [key, val] of value.value.props.entries()) {
        // add indendation
        const valStr = stringOfValue(val)
          .split("\n")
          .map((s, i) => (i > 0 ? `  ${s}` : s))
          .join("\n");
        str += `  ${key}: ${valStr},\n`;
      }
      str += ")";
      return str;
    }

    case "NamedRecordDefinitionGroupInstance": {
      return `[classes: ${value.value.ast.identifier.value} (${value.value.ast.namedRecordDefinitions.length} classes)]`;
    }

    case "AnonymousFunctionInstance": {
      return "<anonymous function>";
    }

    case "empty": {
      return "null";
    }

    default:
      throw exhaustive(kind);
  }
}

function resolveTypeIdentifier(
  identifier: LangType["TypeIdentifier"] | LangType["NestedTypeIdentifier"],
  context: Context
) {
  // TODO: implement
  const { kind } = identifier;
  switch (kind) {
    case "TypeIdentifier": {
      return identifier.value;
    }

    case "NestedTypeIdentifier": {
      return nullthrows(
        identifier.path[identifier.path.length - 1]?.value,
        "parser error: parser should ensure NestedTypeIdentifier is never of length < 2"
      );
    }
  }
}
