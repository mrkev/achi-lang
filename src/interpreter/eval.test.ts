import { Lang } from "../parser/parser";
import { Context } from "./Context";
import { evaluateExpression } from "./evaluateExpression";
import { interpret } from "./interpreter";
import { System } from "./System";

function logsFor(script: string) {
  const system = new System();
  interpret(script, system);
  return system.console._log;
}

test("simple.1", () => {
  const logs = logsFor(`#log "hi"`);
  expect(logs).toMatchSnapshot();
});

test("if.1", () => {
  const logs = logsFor(`
  if (1) {
    #log "pass"
  }
  `);
  expect(logs).toMatchSnapshot();
});

test("class.1", () => {
  const logs = logsFor(`
  class Point(x: number, y: number)
  `);
  expect(logs).toMatchSnapshot();
});

test("func.1", () => {
  const logs = logsFor(`
  function foo matches Card {
    case King(): { }
  };
  #log foo;
  `);
  expect(logs).toMatchSnapshot();
});

test("evaluateExpression.NamedRecordLiteral", () => {
  // const context = Context.create();
  // context.values().set("Point");
  // evaluateExpression();
});
