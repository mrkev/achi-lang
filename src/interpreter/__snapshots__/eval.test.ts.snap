// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`class.1 1`] = `Array []`;

exports[`func.1 1`] = `
Array [
  "[Function: foo]",
]
`;

exports[`if.1 1`] = `
Array [
  "pass",
]
`;

exports[`simple.1 1`] = `
Array [
  "hi",
]
`;
